### PR TITLE
[ShardedDDP] Sync buffers + small cleanup

### LINKED ISSUE #68: Support broadcast_buffers in OssDdp
## 🚀 Feature
We should add support for the `broadcast_buffers` flag to OssDdp.

## Motivation
Distributed training with BatchNorm requires it. We removed it from the fairseq implementation because it slows things down a bit, but for the generalized implementation here we should add it back (as a configurable option).

## Additional context
See documentation for `broadcast_buffers` in the main DDP module: https://pytorch.org/docs/master/generated/torch.nn.parallel.DistributedDataParallel.html

### PR DESCRIPTION
**Pull request by blefaudeux**
# Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
- Fixes #68, adding a 'broadcast buffers' option.
- Removes unused code in ShardedDDP, check for buffers that were just set

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by blefaudeux** [comment]
Long lived branch so rebased a couple of times against master, most of the commits listed above are unrelated

File: tests/nn/data_parallel/test_sharded_ddp.py [hunk 1]
```diff
@@ -58,10 +66,9 @@ def run_one_step(rank, world_size, backend, device, temp_file_name):
             if param.requires_grad:
                 assert param.grad.abs().sum().item() > 0.0, "The reduce step should have populated all the gradients"
 
-    # Check that the optimization process makes sense (ie. loss goes down for the same data)
-    optimizer.step()
-    new_eval = ddp(input_tensor).abs().sum() / input_tensor.numel()
-    # assert new_eval.item() < output.item()
```
Comment (blefaudeux) [inline tests/nn/data_parallel/test_sharded_ddp.py:64]: this would be LR and random values dependent, not a very good test so it was not actually being used

File: fairscale/nn/data_parallel/sharded_ddp.py [hunk 2]
```diff
@@ -118,52 +126,36 @@ def reduce(self) -> None:
         """
         assert self.module.training, "Cannot call reduce in eval"
 
-        def reduce_params(params: List[Parameter], params_rank: int) -> None:
```
Comment (blefaudeux) [inline fairscale/nn/data_parallel/sharded_ddp.py:121]: some of these cases were never used here, I assumed that they came from a copy-paste where this function was more generic than just reducing grads

**Comment by blefaudeux** [comment]
ping reviewers, @msbaines @min-xu-ai

File: fairscale/nn/data_parallel/sharded_ddp.py [hunk 3]
```diff
@@ -195,22 +187,33 @@ def reduction_fn() -> None:
...
+                    reduce_grads(buffered_params, cast(int, param_rank))
 
         reduction_fn()
+
+    def _sync_buffers(self) -> None:
+        _ = list(
+            map(
+                lambda x: x.wait(),
+                map(
+                    lambda x: dist.broadcast(x, self.authoritative_rank, self.process_group, async_op=True),
```
Comment (min-xu-ai) [inline fairscale/nn/data_parallel/sharded_ddp.py:215]: does it make sense to combine some buffers and reduce the total number of broadcast, like what we do above for reduce_grad?

File: fairscale/nn/data_parallel/sharded_ddp.py (hunk 2 above)
Comment (min-xu-ai) [reply fairscale/nn/data_parallel/sharded_ddp.py:121]: yeah, it was me mis-named it before.

**Review by min-xu-ai — APPROVED** [review]
Looks good. Feel free to add a TODO to address my comments on potential perf optimization. sorry for the delay.

File: fairscale/nn/data_parallel/sharded_ddp.py (hunk 3 above)
Comment (blefaudeux) [reply fairscale/nn/data_parallel/sharded_ddp.py:215]: yes, I think it would be an interesting option, I have an old branch doing that which I should revive. There's a trade-off though because there are copies in that case, in practice since the broadcast became async I saw a big speed bump and wondered whether this bucketing strategy is still to be followed ? I'll add a TODO, good idea
